provider/reprovider.go

```diff
@@ -236,208 +302,223 @@ func initialReprovideDelay(duration time.Duration) Option {
-				if err := s.ds.Put(s.ctx, lastReprovideKey, storeTime(s.lastRun)); err != nil {
-					log.Errorf("could not store last reprovide time: %v", err)
-				}
-				if err := s.ds.Sync(s.ctx, lastReprovideKey); err != nil {
-					log.Errorf("could not perform sync of last reprovide time: %v", err)
-				}
-			} else {
-				s.statLk.Unlock()
-			}
-
-			s.throughputDurationSum += dur
-			s.throughputProvideCurrentCount += uint(len(keys))
-			if s.throughputCallback != nil && s.throughputProvideCurrentCount >= s.throughputMinimumProvides {
-				if more := s.throughputCallback(performedReprovide, complete, s.throughputProvideCurrentCount, s.throughputDurationSum); !more {
-					s.throughputCallback = nil
-				}
-				s.throughputProvideCurrentCount = 0
-				s.throughputDurationSum = 0
+			// Deduplicate recently provided cids
+			if found, _ := s.instantProvideDeduplicator.ContainsOrAdd(c, struct{}{}); found {
+				continue
 			}
+			provideOperation(s.ctx, c)
+		case <-s.ctx.Done():
+			return
 		}
-	}()
+	}
+}
 
-	// don't start reprovide scheduling if reprovides are disabled (reprovideInterval == 0)
-	if s.reprovideInterval == 0 {
+func (s *reprovider) reprovideSchedulingWorker() {
+	defer s.closewg.Done()
+
+	// read last reprovide time written to the datastore, and schedule the
+	// first reprovide to happen reprovideInterval after that
+	firstReprovideDelay := s.initalReprovideDelay
+	lastReprovide, err := s.getLastReprovideTime()
+	if err == nil && time.Since(lastReprovide) < s.reprovideInterval-s.initalReprovideDelay {
+		firstReprovideDelay = time.Until(lastReprovide.Add(s.reprovideInterval))
+	}
+	firstReprovideTimer := time.NewTimer(firstReprovideDelay)
+
+	select {
+	case <-firstReprovideTimer.C:
+	case <-s.ctx.Done():
 		return
 	}
 
-	s.closewg.Add(1)
-	go func() {
-		// reprovides scheduling worker
-		defer s.closewg.Done()
-
-		// read last reprovide time written to the datastore, and schedule the
-		// first reprovide to happen reprovideInterval after that
-		firstReprovideDelay := s.initalReprovideDelay
-		lastReprovide, err := s.getLastReprovideTime()
-		if err == nil && time.Since(lastReprovide) < s.reprovideInterval-s.initalReprovideDelay {
-			firstReprovideDelay = time.Until(lastReprovide.Add(s.reprovideInterval))
-		}
-		firstReprovideTimer := time.NewTimer(firstReprovideDelay)
+	// after the first reprovide, schedule periodical reprovides
+	nextReprovideTicker := time.NewTicker(s.reprovideInterval)
 
+	for {
+		err := s.Reprovide(context.Background())
+		if err != nil {
+			if s.ctx.Err() != nil {
+				return
+			}
+			log.Errorf("failed to reprovide: %s", err)
+		}
 		select {
-		case <-firstReprovideTimer.C:
+		case <-nextReprovideTicker.C:
 		case <-s.ctx.Done():
 			return
 		}
+	}
+}
 
-		// after the first reprovide, schedule periodical reprovides
-		nextReprovideTicker := time.NewTicker(s.reprovideInterval)
+func (s *reprovider) run() {
+	s.closewg.Add(1)
+	if s.batchProvides {
+		go s.batchProvideWorker()
```

Comment:
As noted in overall comments in the PR, I do not think there is any remaining purpose for batched provides. Better to adjust the worker pool size.

Reply:
I will remove batched provides as discussed during triage.